tests/migrate_signals/test_operation_signals.py

```diff
@@ -0,0 +1,196 @@
+from io import StringIO
+from unittest.mock import patch
+
+from django.core import management
+from django.db import connection, transaction
+from django.db.migrations import Migration
+from django.db.migrations.operations import CreateModel, RunPython
+from django.db.migrations.recorder import MigrationRecorder
+from django.db.models import signals
+from django.test import TransactionTestCase, override_settings
+
+from .operations import Operation1, Operation2, Operation3
+
+
+class Receiver:
+    def __init__(self, operation):
+        self.sender = operation
+        signals.post_operation.connect(self, sender=operation)
+
+
+class InjectOperation1Receiver(Receiver):
+    def __call__(self, signal, sender, **kwargs):
+        return [Operation1()]
+
+
+class InjectOperation2Receiver(Receiver):
+    def __call__(self, signal, sender, **kwargs):
+        return [Operation2()]
+
+
+class InjectOperation1And2Receiver(Receiver):
+    def __call__(self, signal, sender, **kwargs):
+        return [Operation1(), Operation2()]
+
+
+class InjectOperation3Receiver(Receiver):
+    def __call__(self, signal, sender, **kwargs):
+        return [Operation3()]
+
+
+@override_settings(
+    MIGRATION_MODULES={"migrate_signals": "migrate_signals.custom_migrations"}
+)
+class PostOperationSignalTests(TransactionTestCase):
+
+    available_apps = ["migrate_signals"]
+
+    def setUp(self):
+        self.receivers = []
+
+    def tearDown(self):
+        # Disconnect all the receivers
+        for receiver in self.receivers:
+            signals.post_operation.disconnect(receiver, sender=receiver.sender)
+
+        # Reset applied-migrations state.
+        recorder = MigrationRecorder(connection)
+        recorder.migration_qs.filter(app='migrate_signals').delete()
+
+        # Drop table if created
+        table_name = 'migrate_signals_signal'
+        with connection.cursor() as cursor:
+            table_names = connection.introspection.table_names(cursor)
+            if table_name in table_names:
+                connection.disable_constraint_checking()
+                sql_delete_table = connection.schema_editor().sql_delete_table
+                with transaction.atomic():
+                    cursor.execute(sql_delete_table % {
+                        "table": connection.ops.quote_name(table_name),
+                    })
```

Comment:
I agree that it could be simplified, I just re-used a mixed bag of things I found in other tests to get a first set of tests running. Thanks for the pointers, I'll play with the `executor.apply_migration` and see where it takes me.

Reply:
@charettes I updated the tests to use your suggestion, please, let me know if it is what you had in mind and if you think about tests I should add.